### ISSUE TITLE
GAlib fitness indiviadual choice

### ISSUE THREAD

**Issue by mathius** [crocs-muni/EACirc#71]
Which individuals are chosen, if mutated have the same value?

**Comment by Bender250** [comment]
TL DR: it is random, for both mutation and crossover (which is done simultaneously).

There is the way to get this information. If you thing, that my interpretation seems legit, please, close the issue.
//   Evolve a new generation of genomes.  A steady-state GA has no 'old'
// and 'new' populations - we pick from the current population and replace its
// members with the new ones we create.  We replace the worst members of the
// preceeding population.  If a genome in the tmp population is worse than
// one in the main population, the genome in the main population will be
// replaced regardless of its better score.
GASteadyStateGA::step()
...
-- first part of the function creates new individuals:
// Generate the individuals in the temporary population from individuals in 
// the main population.
-- there is stored either mutated, or unchanged individual from previous gen (in case of enabled crossover -- either parents, or pair of new crossed children (not both))
...
// Replace the worst genomes in the main population with all of the individuals
// we just created.  Notice that we invoke the population's add member with a
// genome pointer rather than reference.  This way we don't force a clone of
// the genome - we just let the population take over.  Then we take it back by
// doing a remove then a replace in the tmp population.
-- here, they add all new individuals in the population (they are added on the end of population array)
-- whole population is resolved in GAPopulation::scale (scaling the population)
// the individuals in tmpPop are all owned by pop, but tmpPop does not know 
// that.  so we use replace to take the individuals from the pop and stick 
// them back into tmpPop
  for(i=0; i<tmpPop->size(); i++)
    tmpPop->replace(pop->remove(GAPopulation::WORST, GAPopulation::SCALED), i);
-- remove will sort the individuals (basic wikipedia quicksort) and removes the worst -> **because quicksort is not stable sort, the removed individual is now random (based on fitness of all individuals)**.


OMG, there were case sensitive variable names, overloaded functions with so different purpose, own macros for basic stuff like max, but worse - unfixed things, possible bugs places :(. And also the efficiency of the choosing algorithm is horrible. The population is sorted replacementSize times (they cannot sort it just once and than remove the rest of unused population (of course, that the population is whole the time almost sorted, so the complexity of (recursive) quicksort is qudratic)).

**Comment by mathius** [comment]
Well, that's what the world means, when people say "legacy code". You should show it to Jirka so that he values my code more :D.
Thanks for the research, I'm closing the issue.
